util/syspolicy/setting/errors.go

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package setting
+
+import "errors"
+
+var (
+	// ErrNotConfigured is returned when the requested policy setting is not configured.
+	ErrNotConfigured = errors.New("not configured")
+	// ErrTypeMismatch is returned when there's a type mismatch between the actual type
+	// of the setting value and the expected type.
+	ErrTypeMismatch = errors.New("type mismatch")
+	// ErrNoSuchKey is returned by [DefinitionOf] when no policy setting
+	// has been registered with the specified key.
+	//
+	// Until 2024-08-02, this error was also returned by a [Handler] when the specified
+	// key did not have a value set. While the package maintains compatibility with this
+	// usage of ErrNoSuchKey, it is recommended to return [ErrNotConfigured] from newer
+	// [source.Store] implementations.
+	ErrNoSuchKey = errors.New("no such key")
+)
+
+// Error is an error when reading or parsing a policy setting.
+type Error struct {
+	text string
+}
+
+// NewError returns a [Error] with the specified error message.
+func NewError(text string) *Error {
+	return &Error{text}
+}
+
+// WrapError returns an [Error] with the text of the specified error,
+// or nil if err is nil, [ErrNotConfigured], or [ErrNoSuchKey].
+func WrapError(err error) *Error {
```

Comment:
The `Error` type is used to marshal errors when a policy setting is sent on the wire. Currently, we only marshal error text so that it can be used for logging/displaying.
Eventually, we might also marshal/unmarshal type information for a subset of platform-agnostic error types.

`ErrNotConfigured` and `ErrNoSuchKey` are not really errors: they're a way for a `Store`/`Handler` to report that a policy setting is not configured. When marshaling, an unconfigured policy setting should be omitted from the JSON output rather than marshaled as if it had an error. A not configured policy setting can can be represented in different ways within the `syspolicy` packages, but never with a non-nil `*Error`.

That being said, I don't particularly like `WrapError` either and would appreciate a better name recommendation.

Reply:
Hmm.

One idea:

Rename the struct `ErrorText` and the func `NewErrorText`? Then add basically what you wrote above to the docstring for `type ErrorText`? 😄 

You could also get a similar effect by storing the `error` and call `.String()` on that when marshalling. Upon unmarshal, the text could be put in `errors.New`. The only hesitation I have on that is then `errors.Is` will work on the instance before marshalling but won't work after unmarshalling, which is fine if used carefully but could also lead to confusing bugs.

Reply:
Yeah, the confusing bugs is the exact reason why I decided not to store `error`.
I would prefer not to have `errors.Is` / `errors.As` that work differently depending on where the policy setting came from.

But renaming and improving the doc is reasonable.